src/index/ivf_raft/ivf_raft.cuh

```diff
@@ -45,103 +49,111 @@
+    }
+    auto
+    ids() {
+        return ids_.view();
+    }
+    auto
+    dists() {
+        return dists_.view();
+    }
+    auto
+    ids_data() {
+        return ids_.data_handle();
+    }
+    auto
+    dists_data() {
+        return dists_.data_handle();
+    }
+
+    auto
+    rows() const {
+        return ids_.extent(0);
+    }
+    auto
+    k() const {
+        return ids_.extent(1);
+    }
+
+ private:
+    raft::device_matrix<std::int64_t, std::int64_t> ids_;
+    raft::device_matrix<float, std::int64_t> dists_;
+};
+
+auto constexpr static const MAX_IVF_FLAT_K = 256;
 
 __global__ void
-prescan(int32_t* g_odata, int32_t* g_idata, int n) {
-    extern __shared__ int32_t temp[];
-    int tx = blockIdx.x * blockDim.x + threadIdx.x;
-    int ty = blockIdx.y * blockDim.y + threadIdx.y;
-    int offset = 1;
-    temp[2 * tx] = g_idata[2 * tx + ty * n];
-    temp[2 * tx + 1] = g_idata[2 * tx + 1 + ty * n];
-    for (int d = n >> 1; d > 0; d >>= 1) {
+postprocess_device_results(bool* enough_valid, int64_t* ids, float* dists, int64_t rows, int k, int target_k,
+                           DeviceBitsetView bitset) {
+    __shared__ int invalid_count[1];
+    for (auto row_index = blockIdx.x; row_index < rows; row_index += gridDim.x) {
+        if (threadIdx.x == 0) {
+            invalid_count[0] = 0;
+        }
         __syncthreads();
-        if (tx < d) {
-            int ai = offset * (2 * tx + 1) - 1;
-            int bi = offset * (2 * tx + 2) - 1;
-            temp[bi] += temp[ai];
+        // First, replace all invalid IDs with -1
+        for (auto col_index = threadIdx.x; col_index < k; col_index += blockDim.x) {
+            auto elem_index = row_index * k + col_index;
+            auto cur_id = ids[elem_index];
+            auto invalid_id = bitset.test(cur_id);
+            if (invalid_id) {  // TODO(wphicks): assuming the branch is worth it here;
+                               // should analyze perf.
+                atomicAdd_block(invalid_count, 1);
+            }
+            invalid_id |= cur_id == std::numeric_limits<int64_t>::max();
+            ids[elem_index] = int(invalid_id) * -1 + int(!invalid_id) * ids[elem_index];
         }
-        offset *= 2;
-    }
-    if (tx == 0) {
-        temp[n - 1] = 0;
-    }
-    for (int d = 1; d < n; d *= 2) {
-        offset >>= 1;
         __syncthreads();
-        if (tx < d) {
-            int ai = offset * (2 * tx + 1) - 1;
-            int bi = offset * (2 * tx + 2) - 1;
-            int32_t t = temp[ai];
-            temp[ai] = temp[bi];
-            temp[bi] += t;
+
+        // Now move all valid results to the front of the row
+        auto cur_valid_index = row_index * k;
+        for (auto col_index = 0; col_index < k; ++col_index) {
+            auto elem_index = row_index * k + col_index;
+            // Just do one row per block for now; can improve this later
+            if (ids[elem_index] != -1 && threadIdx.x == 0) {
```

Comment:
With a batched block-level cumsum (treating invalid as 0's), you might be able to get each thread in the block independently writing to their assigned output locations.

Reply:
I realized my prior comment wasn't explained very well. I just mean treating each thread as a "k" position where it stores 0 if invalid and 1 if valid and doing a cumsum of the resulting mask.

Reply:
Oh that's a great thought! Do the block-level cumsum and then that gives us non-overlapping pairs of input/output locations that each thread can work on swapping. I'm going to try to get some perf numbers on this tomorrow, and if they're weak (and this is a hotspot), that will be one of the first things I try. Very nice idea!

Reply:
Yeah I was about to say if the perf isn't bottlenecked here then there might not even be a point in doing this at all. The idea just popped up as I was looking over the loop in the kernel.

Reply:
> With a batched block-level cumsum (treating invalid as 0's), you might be able to get each thread in the block independently writing to their assigned output locations. Not super important for the first iteration here but just an idea for future optimizations.

It seems that our previous implementation was along these linesi